robot/control/main.cpp

```diff
@@ -145,51 +195,12 @@ int main() {
     moduleList.emplace_back(curTime, KickerModule::period,        KickerModule::runtime,        &kicker);
     moduleList.emplace_back(curTime, BatteryModule::period,       BatteryModule::runtime,       &battery);
     moduleList.emplace_back(curTime, RotaryDialModule::period,    RotaryDialModule::runtime,    &dial);
-    moduleList.emplace_back(curTime, LEDModule::period,           LEDModule::runtime,           &led);
+    moduleList.emplace_back(curTime, LEDModule::period,           LEDModule::runtime,           led.get());
 
-    while (true) {
-        // Must do all timing in systick
-        // Base time we use is in systick, and if you convert from systick
-        // to us and try to compare timepoints, the value will "overflow" at
-        // some random point which messes up the `Current - Start` magic
-        // that automatically deals with delta time using the math behind overflows
-        uint32_t loopStartTime = DWT_GetTick();
-        uint32_t loopEndTime = DWT_GetTick() + SUPER_LOOP_PERIOD*DWT_SysTick_To_us();
+    osThreadDef(SuperLoopTask, superLoopTask, osPriorityAboveNormal, 1, 1 << 15);
+    osThreadCreate(osThread(SuperLoopTask), NULL);
 
-        for (unsigned int i = 0; i < moduleList.size(); i++) {
-            MODULE_META_DATA& module = moduleList.at(i);
-            uint32_t currentTime = DWT_GetTick();
+    osKernelStart();
 
-            // Check if we should run the module
-            //      It's time to run it and
-            //      we have enough time in the super loop
-            //
-            // Subtraction allows for rollover compensation
-            // then convertion to signed allows simple comparison
-            if (static_cast<int32_t>(currentTime - module.nextRunTime) >= 0 &&
-                static_cast<int32_t>(loopEndTime - currentTime) >= module.moduleRunTime) {
-                
-                // todo change to loop start time
-                module.lastRunTime = loopStartTime;
-                module.nextRunTime = loopStartTime + module.modulePeriod;
-
-                module.module->entry();
-            }
-
-            // Check if we missed a module X times in a row
-            if (static_cast<int32_t>(currentTime - module.nextRunTime) > MAX_MISS_CNT*module.moduleRunTime) {
-                //printf("WARNING: Missed module #%d run %d times in a row\r\n", i+1, MAX_MISS_CNT);
-                led.missedModuleRun();
-            }
-        }
-
-        int32_t elapsed = DWT_GetTick() - loopStartTime;
-        if (elapsed < static_cast<int32_t>(SUPER_LOOP_PERIOD*DWT_SysTick_To_us())) {
-            DWT_Delay_Sys(SUPER_LOOP_PERIOD*DWT_SysTick_To_us() - elapsed);
-        } else {
-            //printf("WARNING: Overran super loop time\r\n");
-            led.missedSuperLoop();
-        }
-        
-    }
+    for (;;);
```

Comment:
I always like throwing a yield in here

Reply:
This should be unreachable.

Reply:
How so? osKernalStart returns and continues execution in main base on my reading of the following note.

Note
When the CMSIS-RTOS starts thread execution with the function main this function resumes thread switching. The main thread will continue executing after osKernelStart.

Reply:
Huh. That is different from the documentation that I've seen...I think that note may be referring to when you actually have the initialization sequence in a Task, i.e. main is a task, which we're not doing.

From https://www.keil.com/pack/doc/CMSIS/RTOS2/html/theory_of_operation.html:
> Start the RTOS scheduler using osKernelStart which also configures the system tick timer and initializes RTOS specific interrupts. This function does not return in case of successful execution. Therefore, any application code after osKernelStart will not be executed.

Reply:
Oof. I was looking at the RTOS1 docs not RTOS2

Reply:
Based on the `osStatus` return type on `osKernalInitialize`, I think this actually might be RTOS v1. In v2 the return type is now `osStatus_t`. `osThreadDef` uses a param struct and `osThreadCreate` is called `osThreadNew` in v2.

[Changelog](https://www.keil.com/pack/doc/CMSIS/RTOS2/html/os2MigrationFunctions.html)

In the [migration guide](https://www.keil.com/pack/doc/CMSIS/RTOS2/html/os2MigrationGuide.html), it mentions that main is actually it's own thread, which might make that note applicable again. In v2 it specifies that it will not return, but v1 does not have the same description so I think the behavior is different.

Also, it very much make sounds like `osKernelInitialize` and `osKernelStart` aren't fully required, but highly suggested if you need to set things up without switching threads accidentally. Since the [default define](https://www.keil.com/pack/doc/CMSIS/RTOS2/html/os2MigrationGuide.html) is for main thread to be on, calling `osKernelInitialize` pauses thread switching until `osKernelStart` is called again.